### ISSUE TITLE
"WebAssemblyValueType 7 not recognized" with less-optimized WASM

### ISSUE THREAD

**Issue by arktronic** [RyanLamansky/dotnet-webassembly#35]
Hello,

Compiling `Examples/RunExisting/HelloWorld.c` using `-O0`, `-O1`, or `-Os` causes the following error on `Compile.FromBinary`:

```
ArgumentOutOfRangeException: WebAssemblyValueType 7 not recognized. (Parameter 'valueType')
```

The issue does not occur when using `-O2` or `-O3`. Is this expected? If so, could it be explicitly documented somewhere?

Environment:
- WebAssembly v1.1.0 via NuGet
- Latest emcc:
    emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 2.0.21-git (002374041ab769340e824aa3ba549170e6728d91)
    clang version 13.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project 6e1c1dac4c72cc57f4cd2bc8554e8ac9f2f50b6e)
    Target: wasm32-unknown-emscripten
    Thread model: posix

Thank you!

(As an aside, compiling with `-O3 -s SIDE_MODULE=1` requires two `int`s to be passed to `main()` instead of void/nothing, which was a bit surprising. I'm probably missing another `emcc` directive.)

**Comment by RyanLamansky** [comment]
You're not the first person to get an error like this. Unfortunately, I haven't seen any documentation about what type 7 is supposed to be--it's not in the core WASM spec--so I don't have any way of supporting it.

**Comment by arktronic** [comment]
Interesting. Could this possibly be a parsing error where something is interpreted as value type 7, while it might not even be a value type in the first place?

Here's the WAT (via [wabt](https://webassembly.github.io/wabt/demo/wasm2wat/)) from the `-Os` version of `HelloWorld.c`, which suffers from the type 7 issue:

```wat
(module
  (type $t0 (func))
  (type $t1 (func (param i32)))
  (type $t2 (func (result i32)))
  (type $t3 (func (param i32) (result i32)))
  (type $t4 (func (param i32 i32) (result i32)))
  (import "env" "sayc" (func $env.sayc (type $t1)))
  (import "env" "__memory_base" (global $env.__memory_base i32))
  (import "env" "memory" (memory $env.memory 1))
  (func $__wasm_call_ctors (export "__wasm_call_ctors") (export "__wasm_apply_data_relocs") (type $t0)
    (nop))
  (func $say_something (export "say_something") (type $t3) (param $p0 i32) (result i32)
    (local $l1 i32) (local $l2 i32) (local $l3 i32)
    (local.set $l3
      (local.get $p0))
    (if $I0
      (local.tee $l1
        (i32.load8_u
          (local.get $p0)))
      (then
        (local.set $l2
          (local.get $p0))
        (loop $L1
          (call $env.sayc
            (i32.shr_s
              (i32.shl
                (local.get $l1)
                (i32.const 24))
              (i32.const 24)))
          (local.set $l1
            (i32.load8_u offset=1
              (local.get $l2)))
          (local.set $l2
            (local.tee $l3
              (i32.add
                (local.get $l2)
                (i32.const 1))))
          (br_if $L1
            (local.get $l1)))))
    (i32.sub
      (local.get $l3)
      (local.get $p0)))
  (func $__original_main (export "__original_main") (type $t2) (result i32)
    (drop
      (call $say_something
        (global.get $env.__memory_base)))
    (i32.const 0))
  (func $main (export "main") (type $t4) (param $p0 i32) (param $p1 i32) (result i32)
    (call $__original_main))
  (global $__dso_handle (export "__dso_handle") i32 (i32.const 0))
  (data $d0 (global.get $env.__memory_base) "Hello World (from WASM)\0a\00"))
```

Can this help in determining the root cause?

**Comment by RyanLamansky** [comment]
I don't currently have WASM compiler chain set up so it's actually easier for me to troubleshoot a raw WASM, although having the WAT could help explain where things get weird once I've isolated it in the resulting binary.

**Comment by arktronic** [comment]
Using the other wabt tool demo, [wat2wasm](https://webassembly.github.io/wabt/demo/wat2wasm/), with the above WAT (paste it in and then click Download) produces a raw WASM file that exhibits the issue.

**Comment by munik** [comment]
Thank you for posting the WAT @arktronic! I installed the same version of emscripten but I was having trouble replicating the issue. (As an aside, I'd be interested in knowing the command you used to compile that.)

This looks like it might be due to having both the global import `$env.__memory_base` and declared global `$__dso_handle`.

[Compile.cs:556](https://github.com/RyanLamansky/dotnet-webassembly/blob/7338583be9f428642948fca1d761f9aa8a4f7bac/WebAssembly/Runtime/Compile.cs#L556) sets `globals` to `importGlobals.ToArray()` and then [Compile.cs:738](https://github.com/RyanLamansky/dotnet-webassembly/blob/7338583be9f428642948fca1d761f9aa8a4f7bac/WebAssembly/Runtime/Compile.cs#L738) iterates `globals.Count` times (which is now 2).

The error appears to be occurring during the second iteration when it calls `ReadVarInt7()` to get the content type. But since there aren't actually two globals to read in the globals section, this is whatever data is adjacent to that, not the content type.

If I'm reading this output correctly, wasm-objdump confirms that there is only one global defined in the global section:
```
Contents of section Global:
000005b: 017f 0041 000b                           ...A..
```

I'm still very new to this codebase, so I want to hear what @RyanLamansky thinks of this analysis and, if it is correct, what the fix is. I would hazard a guess that perhaps Compile.cs:738 should loop while `i < count` rather than `i < globals.Length`?

I went ahead and tested with that change, and it results in a different error. (Probably because I didn't add any imports but am running the original Program.cs under the RunExisting example.)
`WebAssembly.Runtime.ImportException: Missing import for env::__memory_base.`

**Comment by arktronic** [comment]
The commandline was: `emcc HelloWorld.c -Os -o a.out.wasm -s SIDE_MODULE=1`

**Comment by arktronic** [comment]
@munik Out of curiosity, how did you compile it? I'm still unsure as to why my compilation results in `main` requiring two `int` arguments. Was that the case when you tried to compile the C code?

**Comment by RyanLamansky** [comment]
> I'm still very new to this codebase, so I want to hear what @RyanLamansky thinks of this analysis and, if it is correct, what the fix is. I would hazard a guess that perhaps Compile.cs:738 should loop while `i < count` rather than `i < globals.Length`?

That's a good analysis--parts of the code are very old and I may have missed this when I fix some of the counters.

> I went ahead and tested with that change, and it results in a different error. (Probably because I didn't add any imports but am running the original Program.cs under the RunExisting example.)
> `WebAssembly.Runtime.ImportException: Missing import for env::__memory_base.`

Yeah, you'll need to satisfy all imports to be able to run it. If you got this far, it means the WASM fully compiled 🙂

**Comment by munik** [comment]
@arktronic This is embarrassing, but I hadn't gotten as far as the `SIDE_MODULE` flag!

I can confirm that when specifying `-s SIDE_MODULE=1` I also get a main with two i32 params. The rest of it has some slight formatting differences but it's all basically the same. Not sure what's going on there.

**Comment by arktronic** [comment]
Looks like everything is working after satisfying the additional import. Thanks, @munik and @RyanLamansky!

**Comment by RyanLamansky** [comment]
@munik's fix has been deployed to NuGet in 1.2.0 👍
